### PR TITLE
XML Diff in Rook now can ignore nodes provided by user

### LINKED ISSUE #2170: [TASK] XML Diff tests should be able to ignore nodes
--------
Issue Description
--------

**Is your feature request related to a problem? Please describe.**
When testing the difference between two XML nodes, RAVEN should be able to ignore certain nodes provided by the user.

**Describe the solution you'd like**
In runtime or when testing, a user could submit an address (in xpath format) for which node(s) should be ignored when comparing. 

**Describe alternatives you've considered**

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by GabrielSoto-INL**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
close #2170 

##### What are the significant changes in functionality due to this change request?
1) `XMLDiff` now has new parameter for `ignored_nodes` which is an xpath format list of strings for ignoring specified child nodes when comparing XML data
2) `XMLDiff` removes the nodes from the loaded XML data (in-memory) before conducting a comparison study
3) Testing this new functionality in `TestXMLDiffer.py`
4) Some linting corrections to existing code, no change in ( range(len(array)) -> enumerate(array) kind of stuff)

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

### CODE REVIEW

File: rook/XMLDiff.py [hunk 1]
```diff
@@ -335,17 +334,44 @@ def fail_message(*args):
...
+    try:
+      for element in [a_element, b_element]:
+        parentpath = '/'.join(ignored.split('/')[:-1]) # `.//parentNode/childNode[@name:<>]`
+        childpath = ignored.split('/')[-1]             # `/grandchildNode`
+        parentnode = element.find(parentpath)
+        parentnode.remove(parentnode.find(childpath))
+      success = True
+    except Exception as exp:
+      print((f"Could not find XPath {ignored} in elements with exception:{exp}"))
+      return a_element, b_element, False
```
Comment (wangcj05) [inline rook/XMLDiff.py:372]: The return will happen if the node can not be found in the first element. In the situation when the node only contained in the second element, the return will still keep the ignored node the second element. I would suggest to remove ignored node in both elements first, and then return them.

**Review by wangcj05 — CHANGES_REQUESTED** [review]
@GabrielSoto-INL : I have some comments for you to consider

File: rook/XMLDiff.py (hunk 1 above)
Comment (GabrielSoto-INL) [reply rook/XMLDiff.py]: that's a good point, I didn't consider the case where we ignore a node found in only one of the element trees. 

I have removed the try-except for a simpler search. It now only exits out of the method when the ignored node is not found in _either_ of the element trees. 

I have also added a test for this case to `TestXMLDiffer.py`. Thanks for catching that!

File: rook/XMLDiff.py [hunk 2]
```diff
@@ -418,7 +459,7 @@ def diff(self):
               gold_root = alt_gold_root_list[0]
           if not self.__same:
             pass #already failed
-          elif 'unordered' in self.__options.keys() and self.__options['unordered']:
+          elif 'unordered' in list(self.__options) and self.__options['unordered']:
```
Comment (joshua-cogliati-inl) [inline rook/XMLDiff.py:462]: I don't think this needs to be changed, but I am not quite sure why this change was made?  ( `'unordered' in self.__options`  would also have worked, the `list` is redundant.)

File: rook/XMLDiff.py [hunk 3]
```diff
@@ -335,17 +334,51 @@ def fail_message(*args):
...
+      parentnode = tree.find(parentpath)
+      if parentnode is not None:
+        childnode = parentnode.find(childpath)
+        if childnode is not None: #simple boolean check doesn't work?
+          parentnode.remove(childnode)
+          found = True
+      successes.append(found)
+    if sum(successes) == 0:
+      print((f"Given XPath {ignored} not found in either tree."))
+      return a_tree, b_tree, False
```
Comment (wangcj05) [inline rook/XMLDiff.py:378]: @GabrielSoto-INL I still have a question regarding this process. It seems to me user can provided multiple ignored nodes since there is a "for loop at the beginning". If this is the case, there will be an early return when the first ignored node is not found in both tree, and the rest of ignored nodes will be skipped. I think we should check all the ignored nodes and return after that.

**Review by joshua-cogliati-inl — DISMISSED** [review]
Looks good to me.

File: rook/XMLDiff.py (hunk 3 above)
Comment (joshua-cogliati-inl) [reply rook/XMLDiff.py:378]: Hm, should the last if statement just be taken out of the for loop?

Comment (GabrielSoto-INL) [reply rook/XMLDiff.py]: I have fixed the issue (and renamed some of the variables to hopefully be clearer)

File: rook/XMLDiff.py (hunk 2 above)
Comment (GabrielSoto-INL) [reply rook/XMLDiff.py]: I was originally getting a Pylint warning C0201 (https://pylint.readthedocs.io/en/latest/user_guide/messages/convention/consider-iterating-dictionary.html) for iterating through the keys with `.keys()`. but you're right that the `list` is redundant, I applied the suggested fix

**Review by wangcj05 — APPROVED** [review]
changes look good.

**Comment by wangcj05** [comment]
Tests are green and checklist is satisfied.

**Comment by wangcj05** [comment]
@GabrielSoto-INL Thanks for your contribution.
